### PR TITLE
[MBL-223]Call startfile/stopfile on a configurable interval for a configurable duration

### PR DESCRIPTION
**Pull request by Jane-XuJingyi**
You can use `stat_profile` worker flag for CPU Profile with a certain interval.
The flag value should be like `interval&duration`(eg, 60&10). Interval is how much time between two profiles. Duration is how much time each profile lasts for. 
The output files are saved as `worker_nanny\data\workers\UnrealWorker\GDKTestGyms\Saved\Profiling\MemReports\Map-Platform-ProfilesessionTimestamp\workerID-SnapshotTimestamp.ue4stats`.
Profiles during the same profile session will be saved to the same folder.

Test results: https://buildkite.com/improbable/product-research-benchmarks-run/builds/21871

### CODE REVIEW

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ namespace
 	const FString StatProfileCommandLineKey = TEXT("-StatProfile=");
 
 	const FString MemReportFlag = TEXT("mem_report");
-	const FString MemRemportIntervalKey = TEXT("-Interval=");
+	const FString MemRemportIntervalKey = TEXT("-MemReportInterval=");
```
Comment (jayprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:61]: What is this -MemReportInterval?

Comment (Jane-XuJingyi) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:61]: It's a key used in command arguments (eg.-MemReportInterval=30). I renamed this key to make it less confusing.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h [hunk 2]
```diff
@@ -122,6 +122,7 @@ class GDKTESTGYMS_API ABenchmarkGymGameModeBase : public AGameModeBase
 	int32 NumSpawnZones;
 #if	STATS
 	// For stat profile
+	int32 CPUProfileInterval;
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h:125]:
```suggestion
#if STATS
	int32 CPUProfileInterval;
#endif
```
This is unused and unset if `STATS` is not set/defined

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h [hunk 3]
```diff
@@ -156,9 +157,9 @@ class GDKTESTGYMS_API ABenchmarkGymGameModeBase : public AGameModeBase
 	void SetLifetime(int32 Lifetime);
 	int32 GetPlayerControllerCount() const;
 #if	STATS
-	void SetStatTimer(const FString& TimeString);
-	void InitMemReportTimer(const FString& MemReportIntervalString);
+	void InitStatTimer(const FString& CPUProfileString);
 #endif
+	void InitMemReportTimer(const FString& MemReportIntervalString);
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h:162]: This is only implemented when `STATS` is enabled, did you also mean to move the implementation out of an #if block?

Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h:162]: I can only see usage of this inside `STATS` though

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 4]
```diff
@@ -256,24 +256,26 @@ void ABenchmarkGymGameModeBase::Tick(float DeltaSeconds)
...
 #if	STATS
-	if (StatStartFileTimer.HasTimerGoneOff())
+	if (CPUProfileInterval > 0 && StatStartFileTimer.HasTimerGoneOff())
 	{
 		FString Cmd(TEXT("stat startfile"));
 		if (GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking())
 		{
 			USpatialNetDriver* SpatialDriver = Cast<USpatialNetDriver>(GetNetDriver());
-			if (ensure(SpatialDriver != nullptr))
+			if (SpatialDriver != nullptr)
```
Comment (jayprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:265]: why was there an ensure here previously?

Comment (Jane-XuJingyi) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:265]: I guess it's from Jungting's previous code...

**Review by jayprobable — APPROVED** [review]

**Review by siliviali — APPROVED** [review]

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 5]
```diff
@@ -256,24 +256,26 @@ void ABenchmarkGymGameModeBase::Tick(float DeltaSeconds)
...
+				const FString Filename = CreateProfileFilename(InFileName, TEXT(".ue4stats"), true);
+				Cmd.Append(FString::Printf(TEXT(" %s"), *Filename));
 			}
 		}
 		GEngine->Exec(GetWorld(), *Cmd);
-		StatStartFileTimer.SetTimer(999999);
+		StatStartFileTimer.SetTimer(CPUProfileInterval);
 	}
-	if (StatStopFileTimer.HasTimerGoneOff())
+	if (CPUProfileInterval > 0 && StatStopFileTimer.HasTimerGoneOff())
```
Comment (jeffxu-improbable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:275]: I prefer to use
if CPUProfileInternal > 0:
{
    if (StatStartFileTimer.HasTimerGoneOff){}
    if (StatStopFileTimer.HasTimerGoneOff){}
}

Comment (Jane-XuJingyi) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp]: Make sense. Updated.

**Review by jeffxu-improbable — APPROVED** [review]

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 6]
```diff
@@ -786,15 +792,21 @@ void ABenchmarkGymGameModeBase::ReportMigration_Implementation(const FString& Wo
 	}
 }
 #if	STATS
-void ABenchmarkGymGameModeBase::SetStatTimer(const FString& TimeString)
+void ABenchmarkGymGameModeBase::InitStatTimer(const FString& CPUProfileString)
 {
-	FString StartDelayString, DurationString;
-	if (TimeString.Split(TEXT(","), &StartDelayString, &DurationString))
+	FString CPUProfileIntervalString, CPUProfileDurationString;
+	if (CPUProfileString.Split(TEXT("&"), &CPUProfileIntervalString, &CPUProfileDurationString))
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:798]: Could we document this format please? Maybe it's worth having some info in a comment block somewhere describing the options we have available for profiling

**Review by improbable-danny — APPROVED** [review]

**Pull request merged**